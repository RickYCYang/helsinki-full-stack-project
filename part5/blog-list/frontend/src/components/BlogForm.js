import React, { useState } from 'react';
import PropTypes from 'prop-types';

const BlogForm = ({ createBlog }) => {
  const [title, setTitle] = useState('');
  const [author, setAuthor] = useState('');
  const [url, setUrl] = useState('');

  const addBlog = async (event) => {
    event.preventDefault();
    await createBlog({
      title,
      author,
      url,
    });
    setTitle('');
    setAuthor('');
    setUrl('');
  };

  return (
    <div>
      <h2>create new</h2>
      <form onSubmit={addBlog}>
        <div>
          title:
          <input
            type="text"
            id="title"
            value={title}
            name="Title"
            placeholder="write here blog title"
            onChange={({ target }) => setTitle(target.value)}
          />
        </div>
        <div>
          author:
          <input
            type="text"
            id="author"
            value={author}
            name="Author"
            placeholder="write here blog author"
            onChange={({ target }) => setAuthor(target.value)}
          />
        </div>
        <div>
          url:
          <input
            type="text"
            id="url"
            value={url}
            name="Url"
            placeholder="write here blog url"
            onChange={({ target }) => setUrl(target.value)}
          />
        </div>
        <button type="submit" id="blog-post-button">
          create
        </button>
      </form>
    </div>
  );
};

BlogForm.propTypes = {
  createBlog: PropTypes.func.isRequired,
};

export default BlogForm;
