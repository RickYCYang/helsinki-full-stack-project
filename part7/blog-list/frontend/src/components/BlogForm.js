import React, { useState } from 'react';
import PropTypes from 'prop-types';
import Button from '@mui/material/Button';

const BlogForm = ({ createBlog }) => {
  const [title, setTitle] = useState('');
  const [author, setAuthor] = useState('');
  const [url, setUrl] = useState('');

  const addBlog = async (event) => {
    event.preventDefault();
    await createBlog({
      title,
      author,
      url,
    });
    setTitle('');
    setAuthor('');
    setUrl('');
  };

  return (
    <div>
      <h2>create new</h2>
      <form onSubmit={addBlog}>
        <div style={{ marginBottom: '12px' }}>
          <div>
            title:
            <input
              type="text"
              id="title"
              value={title}
              name="Title"
              placeholder="write here blog title"
              onChange={({ target }) => setTitle(target.value)}
            />
          </div>
          <div>
            author:
            <input
              type="text"
              id="author"
              value={author}
              name="Author"
              placeholder="write here blog author"
              onChange={({ target }) => setAuthor(target.value)}
            />
          </div>
          <div>
            url:
            <input
              type="text"
              id="url"
              value={url}
              name="Url"
              placeholder="write here blog url"
              onChange={({ target }) => setUrl(target.value)}
            />
          </div>
        </div>
        <Button type="submit" variant="contained" sx={{ marginBottom: '16px' }}>
          create
        </Button>
      </form>
    </div>
  );
};

BlogForm.propTypes = {
  createBlog: PropTypes.func.isRequired,
};

export default BlogForm;
