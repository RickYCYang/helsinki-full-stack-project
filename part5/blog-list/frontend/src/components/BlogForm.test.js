import React from 'react';
import { render, screen, cleanup } from '@testing-library/react';
import '@testing-library/jest-dom/extend-expect';
import BlogForm from './BlogForm';
import userEvent from '@testing-library/user-event';
import { act } from 'react-dom/test-utils';

describe('<BlogForm /> updates parent state and calls onSubmit', () => {
  const createBlog = jest.fn();
  let container;
  beforeEach(() => {
    container = render(<BlogForm createBlog={createBlog} />).container;
  });
  afterEach(() => {
    cleanup();
  });
  test('updates parent state and calls onSubmit', async () => {
    //let [titleInput, authorInput, urlInput] = screen.getAllByRole('textbox');
    const titleInput = screen.getByPlaceholderText('write here blog title');
    const authorInput = screen.getByPlaceholderText('write here blog author');
    const urlInput = screen.getByPlaceholderText('write here blog url');
    const sendButton = screen.getByText('create');

    const title = 'Hello World';
    const author = 'Rick';
    const url = 'www.google.com';

    await act(async () => {
      await userEvent.type(titleInput, title);
      expect(titleInput.value).toBe(title);
      await userEvent.type(authorInput, author);
      expect(authorInput.value).toBe(author);
      await userEvent.type(urlInput, url);
      expect(urlInput.value).toBe(url);
      await userEvent.click(sendButton);
    });

    expect(createBlog).toHaveBeenCalledTimes(1);
    expect(createBlog.mock.calls[0][0].title).toBe(title);
    expect(createBlog.mock.calls[0][0].author).toBe(author);
    expect(createBlog.mock.calls[0][0].url).toBe(url);
    expect(titleInput.value).toBe('');
    expect(authorInput.value).toBe('');
    expect(urlInput.value).toBe('');
  });

  test('snapshot of html elements', () => {
    expect(container).toMatchSnapshot();
  });
});
