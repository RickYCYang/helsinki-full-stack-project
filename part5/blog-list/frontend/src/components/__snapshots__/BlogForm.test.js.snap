// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<BlogForm /> updates parent state and calls onSubmit snapshot of html elements 1`] = `
<div>
  <div>
    <h2>
      create new
    </h2>
    <form>
      <div>
        title:
        <input
          id="title"
          name="Title"
          placeholder="write here blog title"
          type="text"
          value=""
        />
      </div>
      <div>
        author:
        <input
          id="author"
          name="Author"
          placeholder="write here blog author"
          type="text"
          value=""
        />
      </div>
      <div>
        url:
        <input
          id="url"
          name="Url"
          placeholder="write here blog url"
          type="text"
          value=""
        />
      </div>
      <button
        id="blog-post-button"
        type="submit"
      >
        create
      </button>
    </form>
  </div>
</div>
`;
